cjworkbench/forms/signup.py

```diff
@@ -43,3 +43,5 @@ def signup(self, request, user):
         profile, created = UserProfile.objects.get_or_create(user=user)
         profile.get_newsletter = self.cleaned_data["get_newsletter"]
         profile.save()
+        user.user_profile.locale_id = request.locale_id
```

Comment:
`user.user_profile` should be `profile` here.

In fact, this should all be exactly one call. And I think the call should be `update_or_create()`. So how about:

```
UserProfile.objects.update_or_create(
    user=user,
    defaults={"get_newsletter": self.cleaned_data["get_newsletter"], "locale_id": request.locale_id},
)
```

Reply:
For some very weird reason, this always sets the default locale. 
That's why I used `user.user_profile` in the first place: I had tried 
```
profile = UserProfile.objects.update_or_create(user=user)
profile.locale_id = request.locale_id
profile.get_newsletter = self.cleaned_data["get_newsletter"]
profile.save()
```
and it always set the default locale too.

Reply:
"Very weird reason" is a red flag. We should not merge very weird stuff. We should debug, come up with a simple explanation, and put that in a comment.

Here's my theory: we actually create UserProfile in two places:

* `cjworkbench/models/userprofile.py`
* `cjworkbench/forms/signup.py`

... and maybe something along the lines of: allauth is creating the user; and then our custom signal is creating the user _with the default locale_; then the user is signing in and middleware sets `request.locale_id` to user_profile.id; and _then_ we call `signup()`. (This explanation doesn't agree with the facts you've we've laid out; but then, we're in "very weird" territory so please indulge me the brainstorm... :).)

Ideally, we'd track down every call of `django.contrib.auth.models.UserManager.create_user()`. and make sure we create a UserProfile at the same time. (Then we'd nix the signal in `cjworkbench/models/userprofile.py`.) Who calls `create_user()`? I can't find it when I grep Django or Allauth code.

After we figure things out, I'd probably be fine with leaving the user_profile in signup.py. We could read the locale from the request _cookie_ (not the request itself) and write _that_ to the user profile.

Reply:
Hmmm... Turns out it can work as intended when using `profile` instead of `user.user_profile`. Perhaps I had forgotten to add `update_fields` in `profile.save()` when I was experimenting.